src/basic/cgroup-util.c

```diff
@@ -1198,6 +1198,34 @@ int cg_path_get_unit(const char *path, char **ret) {
         return 0;
 }
 
+int cg_path_get_unit_path(const char *path, char **ret) {
+        char *path_copy, *unit_name, *unit_path;
+        size_t unit_name_len;
+
+        assert(path);
+        assert(ret);
+
+        path_copy = strdupa_safe(path);
```

Comment:
instead of copying the string twice (once here via strdupa_alloc() on the stack, and once below via strdup() on the heap). Just allocate it here onto the heap right away.

Reply:
i.e.

```c
_cleanup_free_ char *path_copy = strdup(path);
if (!path_copy)
        return -ENOMEM;

unit_name = (char*) skip_slices(path_copy);
unit_name[strcspn(unit_name, "/")] = 0;

if (!unit_name_is_valid(cg_unescape(unit_name, UNIT_NAME_PLAIN|UNIT_NAME_INSTANCE))
        return -ENXIO;

*ret = TAKE_PTR(path_copy);
return 0;
``